include/trieste/rewrite.h

```diff
@@ -1191,4 +1303,77 @@ namespace trieste
 
     return nodes;
   }
+
+  /// @brief This function can be wrapped around a pattern and effect to debug
+  /// pattern matching. It will count how many tokens have been matched before
+  /// the match failed. The results will be emitted as an error node.
+  ///
+  /// The first argument is the token that the debug should start tracking. To
+  /// investigate why <code>(if)(lit 1:3)(less)</code> doesn't match, the first
+  /// token should be set to <code>If</code> like this:
+  ///
+  /// \code{.diff}
+  ///   {
+  /// +   Dbg(If,
+  ///     T(If) * T(Lit)[Lit] * T(Then) >>
+  ///       [](Match& _) {
+  ///         // [...]
+  ///       }
+  /// +   ),
+  ///   }
+  /// \endcode
+  ///
+  /// The output will contain the number of matched tokens and a list of all
+  /// groups that have been captured until the match fails.
+  ///
+  /// @param dbg_on_success Indicates if successful matches should also emit
+  /// an `err` node with debug information. Setting this to `false` can result
+  /// in the matching on the rewritten tree on consecutive match attempts.
+  inline detail::PatternEffect<Node> Dbg(
+    Token type, detail::PatternEffect<Node> effect, bool dbg_on_success = true)
+  {
+    // Note: This doesn't compile with `auto`. I'm guessing that
+    // `intrusive_ptr` doesn't support downcasting to `PatternPtr`
+    // for the `Pattern` constructor. The hack is to do upcasting
+    // in out custom effect.
+    detail::PatternPtr pattern = intrusive_ptr<detail::DebugMatch>::make(
+      type, effect.first.value, dbg_on_success);
+    auto original_effect = effect.second;
+    return {
+      detail::Pattern(
+        pattern,
+        // We want to accept any token, since the contained pattern could
+        // start with any token, and when want to be invisible until the
+        // start token was found
+        detail::FastPattern::match_pred()),
```

Comment:
I think this could be 
```suggestion
        detail::FastPattern::match_token({type})),
```

Not sure it is worth changing as this is a Debug feature so perf is probably not an issue.

Reply:
If the wrapped pattern can start with `T(type1, type2)` wouldn't `detail::FastPattern::match_token({type1}))` prevent the pattern being matched on `type2`?

I basically tried to make it still work for every token besides the one set in the `Dbg` function

Reply:
Ah, yes you are right